File: ASD/strutture_dati/coda.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>


//Algoritmo FIFO

typedef struct node{
    int e;
    struct node *next;
}node;

typedef struct coda{   
    struct node *start;
    struct node *last;
    
    int len;
}coda;


node* init_node(int value){
    node *nd = malloc(sizeof(node));
    
    nd->e = value;
    nd->next = NULL;

    return nd;
}

coda* init_code(){
    coda *C = malloc(sizeof(coda));

    C->start = NULL;
    C->last = NULL;
    C->len = 0;

    return C;
}

void printer(coda *C){
    printf("[");
    node *nd = C->start;
    while (nd != NULL) {
        printf("%d", nd->e);
        if (nd->next != NULL) {
            printf(", ");
        }
        nd = nd->next;
    }
    printf("]\n");
}

bool isEmpty(coda *C){
    return C->start == NULL;
}


void enqueue(coda *C, int e){
    node *nd = init_node(e);
    if(isEmpty(C)){
        C->start = nd;
    }else{
        C->last->next = nd; 
    }
    C->last = nd;
    C->len++;

}

int dequeue(coda * C){
    node *nd = C->start;

    if(nd == NULL){
        printf("Bru la coda è vuota: ");
        return -1;
    }else{
        int e = C->start->e;
        C->start = C->start->next;
        if(C->start == NULL){
            C->last = NULL;
        }
        free(nd);
        C->len--;
        return e;
    }

}

int first(coda *C){
    return C->start->e;
}

int main(){
    coda *C = init_code();

    enqueue(C, 10);
    enqueue(C, 20);
    enqueue(C, 30);

    printf("Coda dopo gli accodamenti: ");
    printer(C);

    printf("%d\n",dequeue(C));
    printf("%d\n",dequeue(C));
    printf("%d\n",dequeue(C));
    printf("%d\n",dequeue(C));

    printf("Coda dopo le cancellazioni: ");
    printer(C);

    return 0;
}

File: ASD/strutture_dati/coda.py
def isEmpty(C):
    return len(C) == 0

def enqueue(C, elem):
    C.append(elem)

def dequeue(C, current):
    if isEmpty(C):
        return None, current
    
    e = C[current]
    C[current] = None
    current+=1

    if current > len(C)/2:
        new_C = []
        for i in range(current,len(C)):
            new_C.append(C[i])
        C = new_C
        current = 0
    
    return C, e, current

def first(C, current):
    if isEmpty(C):
        return None
        
    return C[current]

#C = []
#current = 0
#
#print(f"La coda è vuota -> {isEmpty(C)}")
#for i in range(10):
#    len_P = enqueue(C,i)
#
#print(f"Coda dopo gli inserimento -> {C}")
#print(f"La coda è vuota -> {isEmpty(C)}")
#
#for i in range(7):
#    C, e, current = dequeue(C, current)
#    print(f"Ho eliminato l'elemento -> {e}") 
#
#print(f"Coda dopo le cancellazione -> {C}")
#
#enqueue(C,10)
#print(f"Coda dopo l'ultimo inserimento -> {C}")
#
#print(f"L'elemento in cima è -> {first(C,current)}")


